# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/BodyStream.cpp
# Commit: ff1de20097f6
# Full Hash: ff1de20097f6aa13cb7e27d827c42e44de4269e4
# Author: Kagami Sascha Rosylight <krosylight@mozilla.com>
# Date: 2023-02-21 21:26:48
# Regressor Bug: 1813011
# File Overlap Count: 1
# Description:
#   Bug 1813011 - Part 3: Remove BodyStream::mState r=smaug
#   
#   1. eInitializing is only used to call MarkAsRead, but it can just be called unconditionally since it's just a boolean setter.
#   2. eClosed is aliased to `!mStreamHolder`, since it becomes null when being closed.
#   
# ==============================================================================

diff -r 68fbd6ac34b2 -r ff1de20097f6 dom/base/BodyStream.cpp
--- a/dom/base/BodyStream.cpp	Tue Feb 21 12:37:38 2023 +0000
+++ b/dom/base/BodyStream.cpp	Tue Feb 21 12:37:38 2023 +0000
@@ -191,16 +191,12 @@
 
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
-  MOZ_DIAGNOSTIC_ASSERT(mState == eInitializing || mState == eInitialized);
+  MOZ_DIAGNOSTIC_ASSERT(!IsClosed());
   MOZ_ASSERT(!mPullPromise);
   mPullPromise = Promise::CreateInfallible(aController.GetParentObject());
 
-  if (mState == eInitializing) {
-    // The stream has been used for the first time.
-    mStreamHolder->MarkAsRead();
-  }
-
-  mState = eInitialized;
+  // Reading the stream, let's mark it as such
+  mStreamHolder->MarkAsRead();
 
   if (!mInputStream) {
     // This is the first use of the stream. Let's convert the
@@ -248,7 +244,7 @@
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
   MOZ_DIAGNOSTIC_ASSERT(mInputStream);
-  MOZ_DIAGNOSTIC_ASSERT(mState == eInitialized);
+  MOZ_DIAGNOSTIC_ASSERT(!IsClosed());
   MOZ_DIAGNOSTIC_ASSERT(mPullPromise->State() ==
                         Promise::PromiseState::Pending);
 
@@ -297,8 +293,8 @@
 void BodyStream::CloseInputAndReleaseObjects() {
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
-  if (mState == eInitializing) {
-    // The stream has been used for the first time.
+  if (mStreamHolder) {
+    // Being closed means someone touched the stream, let's mark it as read.
     mStreamHolder->MarkAsRead();
   }
 
@@ -319,8 +315,7 @@
 BodyStream::BodyStream(nsIGlobalObject* aGlobal,
                        BodyStreamHolder* aStreamHolder,
                        nsIInputStream* aInputStream)
-    : mState(eInitializing),
-      mGlobal(aGlobal),
+    : mGlobal(aGlobal),
       mStreamHolder(aStreamHolder),
       mOwningEventTarget(aGlobal->EventTargetFor(TaskCategory::Other)),
       mOriginalInputStream(aInputStream) {
@@ -337,7 +332,7 @@
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
   // Nothing to do.
-  if (mState == eClosed) {
+  if (IsClosed()) {
     return;
   }
 
@@ -364,8 +359,8 @@
     NS_WARNING_ASSERTION(!rv.Failed(), "Failed to error BodyStream");
   }
 
-  if (mState == eInitializing) {
-    // The stream has been used for the first time.
+  if (mStreamHolder) {
+    // Being errored means someone touched the stream, let's mark it as read.
     mStreamHolder->MarkAsRead();
   }
 
@@ -428,7 +423,7 @@
   mAsyncWaitWorkerRef = nullptr;
 
   // Already closed. We have nothing else to do here.
-  if (mState == eClosed) {
+  if (IsClosed()) {
     return NS_OK;
   }
 
@@ -441,7 +436,6 @@
   AutoEntryScript aes(mGlobal, "fetch body data available");
 
   MOZ_DIAGNOSTIC_ASSERT(mInputStream);
-  MOZ_DIAGNOSTIC_ASSERT(mState == eInitialized);
 
   JSContext* cx = aes.cx();
   ReadableStream* stream = mStreamHolder->GetReadableStreamBody();
@@ -513,7 +507,7 @@
 void BodyStream::Close() {
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
-  if (mState == eClosed) {
+  if (IsClosed()) {
     return;
   }
 
@@ -534,7 +528,7 @@
 void BodyStream::CloseAndReleaseObjects(JSContext* aCx,
                                         ReadableStream* aStream) {
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
-  MOZ_DIAGNOSTIC_ASSERT(mState != eClosed);
+  MOZ_DIAGNOSTIC_ASSERT(!IsClosed());
 
   ReleaseObjects();
 
@@ -548,13 +542,11 @@
 void BodyStream::ReleaseObjects() {
   MOZ_DIAGNOSTIC_ASSERT(mOwningEventTarget->IsOnCurrentThread());
 
-  if (mState == eClosed) {
+  if (IsClosed()) {
     // Already gone. Nothing to do.
     return;
   }
 
-  mState = eClosed;
-
   if (NS_IsMainThread()) {
     nsCOMPtr<nsIObserverService> obs = mozilla::services::GetObserverService();
     if (obs) {